src/cwe_checker_lib/src/analysis/taint/state.rs

```diff
@@ -459,6 +449,154 @@ impl State {
+
+        for (reg, other_taint) in other_register_taint.iter() {
+            match self.register_taint.entry(reg.clone()) {
+                Occupied(mut taint) => {
+                    // FIXME: We create a new owned taint even though we could
+                    // modify in-place.
+                    let new_taint = taint.get().merge(other_taint);
+
+                    taint.insert(new_taint);
+                }
+                Vacant(entry) => {
+                    entry.insert(*other_taint);
+                }
+            }
+        }
+    }
+
+    /// Merges the given pair of abstract identifier and memory object into the
+    /// state.
+    ///
+    /// Equivalent to merging
+    ///     bottom x (`aid` -> (`other_memory_object` + `offset`))
+    /// into the state.
+    fn merge_memory_object_with_offset(
+        &mut self,
+        aid: &AbstractIdentifier,
+        other_memory_object: &MemRegion<Taint>,
+        offset: i64,
+    ) {
+        use hash_map::Entry::*;
+
+        match self.memory_taint.entry(aid.clone()) {
+            Occupied(mut current_memory_object) => {
+                let current_memory_object = current_memory_object.get_mut();
+
+                merge_memory_object_with_offset(current_memory_object, other_memory_object, offset);
+            }
+            Vacant(entry) => {
+                let mut new_memory_object = other_memory_object.clone();
+
+                new_memory_object.add_offset_to_all_indices(offset);
+                entry.insert(new_memory_object);
+            }
+        }
+    }
+
+    /// Deconstructs a `State` into its register and memory taint maps.
+    pub fn into_mem_reg_taint(self) -> (RegisterTaint, MemoryTaint) {
+        (self.register_taint, self.memory_taint)
+    }
+
+    /// Constructs a `State` from register and memory taint maps.
+    pub fn from_mem_reg_taint(register_taint: RegisterTaint, memory_taint: MemoryTaint) -> Self {
+        Self {
+            register_taint,
+            memory_taint,
+        }
+    }
+}
+
+// FIXME: The used algorithm for merging the taints contained in memory
+// regions is unsound when merging taints that intersect only partially.
+// If, for example, in state A `RSP[0:3]` is tainted and in state B
+// `RSP[2:3]` is tainted, A u B will only report `RSP[2:3]` as tainted.
+//
+// For the NULL pointer dereference check, however, this should not have an
+// effect in practice, since these values are usually unsound or a sign of
+// incorrect behavior of the analysed program.
+// FIXME: This looks a lot like it should be a method on `MemRegion<T>`.
+fn merge_memory_object_with_offset(
+    mem_object: &mut MemRegion<Taint>,
+    other_memory_object: &MemRegion<Taint>,
+    offset: i64,
+) {
+    for (index, taint) in other_memory_object.iter() {
+        mem_object.insert_at_byte_index(*taint, *index + offset);
+        // FIXME: Unsound in theory for partially intersecting
+        // taints. Should not matter in practice.
+    }
```

Comment:
Should we also add a warning that this code will become wrong when we allow more complicated taint types in the future (because it does not use the merge function for taints)?